src/charm.py

```diff
@@ -0,0 +1,301 @@
+            return
+        else:
+            self.logger.info(
+                "Resources are missing.  Triggering install to reconcile resources"
+            )
+            self.unit.status = MaintenanceStatus(
+                "Missing kubernetes resources detected - reinstalling"
+            )
+            self._install(event)
+            return
+
+    def _remove(self, _):
+        """Remove charm"""
+        raise NotImplementedError()
+
+    def _create_crds(self, if_exists=None):
+        self.logger.info("Applying manifests for CRDs")
+        objs = self._render_crds()
+        create_all_lightkube_objects(
+            objs, if_exists=if_exists, lightkube_client=self.lightkube_client
+        )
+
+    def _create_rbac(self, if_exists=None):
+        self.logger.info("Applying manifests for RBAC")
+        objs = self._render_rbac()
+        create_all_lightkube_objects(
+            objs, if_exists=if_exists, lightkube_client=self.lightkube_client
+        )
+
+    def _create_controller(self, if_exists=None):
+        self.logger.info("Applying manifests for controller")
+        objs = self._render_controller()
+        create_all_lightkube_objects(
+            objs, if_exists=if_exists, lightkube_client=self.lightkube_client
+        )
+
+    def _render_yaml(self, yaml_filename: [str, Path]):
+        """Returns a list of lightkube k8s objects for a yaml file, rendered in charm context"""
+        context = {
+            "namespace": self.model.name,
+            "metacontroller_image": METACONTROLLER_IMAGE,
+        }
+        with open(self._manifests_file_root / yaml_filename) as f:
+            return codecs.load_all_yaml(f, context=context)
+
+    def _render_crds(self):
+        """Returns a list of lightkube k8s objects for the CRDs, rendered in charm context"""
+        return self._render_yaml("metacontroller-crds-v1.yaml")
+
+    def _render_rbac(self):
+        """Returns a list of lightkube k8s objects for the charm RBAC, rendered in charm context"""
+        return self._render_yaml("metacontroller-rbac.yaml")
+
+    def _render_controller(self):
+        """Returns a list of lightkube k8s objects for the controller, rendered in charm context"""
+        return self._render_yaml("metacontroller.yaml")
+
+    def _render_all_resources(self):
+        """Returns a list of lightkube8s objects for all resources, rendered in charm context"""
+        return self._render_rbac() + self._render_crds() + self._render_controller()
+
+    def _check_deployed_resources(self):
+        """Check the status of all deployed resources, returning True if ok"""
+        expected_resources = self._render_all_resources()
+        found_resources = [None] * len(expected_resources)
+        for i, resource in enumerate(expected_resources):
+            self.logger.info(f"Checking for '{resource.metadata}'")
+            try:
+                found_resources[i] = get_k8s_obj(resource, self.lightkube_client)
```

Comment:
Once again, we could call `lightkube.client.get` here.

Reply:
I find the syntax for lightkube.get just makes it hard to see the intent.  Am I doing something wrong?

This feels awkward to read:
```
something = client.get(type(obj), obj.metadata.name, namespace=obj.metadata.namespace)
```

This feels easier to read (but hides the implementation and adds a few lines of code)
```
something = get_k8s_obj(obj, self.lightkube_client

def get_k8s_obj(obj, client=None):
    if not client:
        client = lightkube.Client()

    return client.get(type(obj), obj.metadata.name, namespace=obj.metadata.namespace)
```

Am I doing something wrong?  It feels like there should be an easier way to `get` an object that I've already defined in lightkube.  

The passing of the client part is also annoying/.  I could move the function into the class to avoid it

Reply:
Nothing fundamentally wrong, I've done it exactly like that. Though I'd say `get` feels natural to me, it looks a lot like `kubectl get <object>/<object name> <namespace>`. I guess my point is, if this helper method is just called once, and it just "wraps" `lightkube.get()` so you don't have to pass the namespace and the obj type, we might as well not have it.

Reply:
I agree with @DnPlas, the syntax for `client.get` might not be the cleanest but I think having an additionall method to abstract that is unnecesarry lets stick with just `lightkube.get()`.